include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+			G2_over_G1 = irr_glsl_ggx_smith_G2_over_G1(NdotL, s.TdotL*s.TdotL, s.BdotL*s.BdotL, NdotL2, NdotV, TdotV2, BdotV2, NdotV2, a2, ay2);
+			ndf_val = irr_glsl_ggx_aniso(TdotH2, BdotH2, NdotH2, a, ay, a2, ay2);
 #endif
-
-		} else
+		} CASE_END
 #endif
 
 #ifdef NDF_BECKMANN
-		if (ndf==NDF_BECKMANN) {
-
+		CASE_BEGIN(ndf, NDF_BECKMANN) {
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction.isotropic, a2);
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(NdotV2, a2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, NdotL2, a2);
+			ndf_val = irr_glsl_beckmann(a2, NdotH2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_DG(currBSDFParams, currInteraction, a, a2, ay, ay2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction, a, a2, ay, ay2);
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, a2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL*s.TdotL, s.BdotL*s.BdotL, NdotL2, a2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, a2, ay2, TdotH2, BdotH2, NdotH2);
 #endif
-
-		} else
+		} CASE_END
 #endif
 
 #ifdef NDF_PHONG
-		if (ndf==NDF_PHONG) {
-
+		CASE_BEGIN(ndf, NDF_PHONG) {
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, n, a2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction.isotropic, a2);
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(NdotV2, a2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, NdotL2, a2);
+			ndf_val = irr_glsl_beckmann(a2, NdotH2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction, a, a2, ay, ay2);
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, a2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, NdotL2, a2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, a2, ay2, TdotH2, BdotH2, NdotH2);
 #endif
-
-		} else
+		} CASE_END
 #endif
-
-#ifndef ONLY_ONE_NDF
+		CASE_OTHERWISE
 		{} //else "empty braces"
-#endif
-	}
+		END_CASES
 
-	uvec3 regs = instr_decodeRegisters(instr);
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
+		bxdf_eval_scalar_part = 0.5 * G2_over_G1 * G1_over_2NdotV * ndf_val; //cos*D*G2/(4*cos*cos)
```

Comment:
why don't you just do `G2_over_G1` here and then mul the thing by `pdf` after you get it.. since the PDF is exactly the other part (`G1_over_2NdotV * ndf_val *0.5` for conductor and the other thing for dielectric refraction) ?

Reply:
new implementation is nice